lib/dpl/provider/lambda.rb

```diff
@@ -20,27 +20,66 @@ def lambda_options
       end
 
       def push_app
-        # Options defined at
-        #   https://docs.aws.amazon.com/sdkforruby/api/Aws/LambdaPreview/Client.html
-        #   https://docs.aws.amazon.com/lambda/latest/dg/API_CreateFunction.html
-        response = lambda.upload_function({
-          function_name:  options[:name]           || option(:function_name),
-          description:    options[:description]    || default_description,
-          timeout:        options[:timeout]        || default_timeout,
-          memory_size:    options[:memory_size]    || deafult_memory_size,
-          role:           option(:role),
-          handler:        handler,
-          function_zip:   function_zip,
-          runtime:        options[:runtime]        || default_runtime,
-          mode:           default_mode,
-        })
-
-        log "Uploaded lambda: #{response.function_name}."
-      rescue ::Aws::LambdaPreview::Errors::ServiceException => exception
+
+        # The original LambdaPreview client supported create/update in one call
+        # To keep compatibility we try to fetch the function and then decide
+        # whether to update the code or create a new function
+
+        begin
+          function_name = options[:name] || option(:function_name)
+
+          response = lambda.get_function({
+            function_name: function_name
+          })
+
+          log "Function #{function_name} already exists, updating."
+
+          # Options defined at
+          #   https://docs.aws.amazon.com/sdkforruby/api/Aws/Lambda/Client.html#update_function_configuration-instance_method
+          response = lambda.update_function_configuration({
+            function_name:  function_name,
+            description:    options[:description]    || default_description,
+            timeout:        options[:timeout]        || default_timeout,
+            memory_size:    options[:memory_size]    || default_memory_size,
+            role:           option(:role),
+            handler:        handler,
+            runtime:        options[:runtime]        || default_runtime,
+          })
+          log "Updated configuration of function: #{response.function_name}."
+
+          # Options defined at
+          #   https://docs.aws.amazon.com/sdkforruby/api/Aws/Lambda/Client.html#update_function_code-instance_method
+          response = lambda.update_function_code({
+             function_name:  options[:name] || option(:function_name),
+             zip_file:       function_zip,
+             publish:        publish,
+          })
+
+          log "Updated code of function: #{response.function_name}."
+        rescue ::Aws::Lambda::Errors::ResourceNotFoundException
```

Comment:
Oh, I see what you are doing now.

Please do not use `rescue` for logical flow that you expect. Check if the function exists first, then deal with it accordingly.